pkg/sysched/sysched.go

```diff
@@ -0,0 +1,506 @@
+	union(sc.HostSyscalls[pod.Status.HostIP], syscall)
+}
+
+func (sc *SySched) addPod(pod *v1.Pod) {
+	nodeIP := pod.Status.HostIP
+	name := pod.Name
+
+	_, ok := sc.HostToPods[nodeIP]
+	if !ok {
+		sc.HostToPods[nodeIP] = make([]*v1.Pod, 0)
+		sc.HostToPods[nodeIP] = append(sc.HostToPods[nodeIP], pod)
+		sc.HostSyscalls[nodeIP] = make(map[string]bool)
+		sc.updateHostSyscalls(pod)
+		return
+	}
+
+	for _, p := range sc.HostToPods[nodeIP] {
+		if p.Name == name {
+			return
+		}
+	}
+
+	sc.HostToPods[nodeIP] = append(sc.HostToPods[nodeIP], pod)
+	sc.updateHostSyscalls(pod)
+
+	return
+}
+
+func (sc *SySched) recomputeHostSyscalls(pods []*v1.Pod) map[string]bool {
+	syscalls := make(map[string]bool)
+
+	for _, p := range pods {
+		syscall := sc.getSyscalls(p)
+		union(syscalls, syscall)
+	}
+
+	return syscalls
+}
+
+func (sc *SySched) removePod(pod *v1.Pod) {
+	nodeIP := pod.Status.HostIP
+
+	_, ok := sc.HostToPods[nodeIP]
+	if !ok {
+		klog.Infof("removePod: Host %s not yet cached", nodeIP)
+
+		return
+	}
+	for i, p := range sc.HostToPods[nodeIP] {
+		if p.Name == pod.Name {
+			sc.HostToPods[nodeIP] = remove(sc.HostToPods[nodeIP], i)
+			sc.HostSyscalls[nodeIP] = sc.recomputeHostSyscalls(sc.HostToPods[nodeIP])
+			c, _ := sc.getHostSyscalls(nodeIP)
+			klog.Info("remaining syscalls: ", c, " ", nodeIP)
+
+			return
+		}
+	}
+
+	return
+}
+
+/*
+func (sc *SySched) podAdded(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	klog.Infof("POD CREATED: %s/%s", pod.Namespace, pod.Name)
+}*/
+
+func (sc *SySched) podUpdated(old, new interface{}) {
+	pod := old.(*v1.Pod)
+	klog.Infof(
+		"POD UPDATED. %s/%s %s",
+		pod.Namespace, pod.Name, pod.Status.Phase,
+	)
+
+	if pod.Status.Phase == v1.PodSucceeded {
+		// finished
+		sc.removePod(pod)
+	} else if pod.Status.Phase == v1.PodPending && pod.Status.HostIP != "" {
+		//klog.Info("pod pending with IP assigned")
+		sc.addPod(pod)
+		//} else if (pod.Status.Phase == v1.PodRunning) {
+		//klog.Info("pod running")
+		//sc.addPod(pod)
+	} else if pod.Status.Phase == v1.PodFailed {
+		if pod.Status.Message != "" {
+			klog.Infof("pod failed: %s", pod.Status.Message)
+		}
+		klog.Infof("pod failed: pod.Status.Message unknown")
+	}
+}
+
+func (sc *SySched) podDeleted(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	klog.Infof("POD DELETED: %s/%s", pod.Namespace, pod.Name)
+	sc.removePod(pod)
+}
+
+// getArgs : returns the arguments for the SySchedArg plugin.
+func getArgs(obj runtime.Object) (*pluginconfig.SySchedArgs, error) {
+	SySchedArgs, ok := obj.(*pluginconfig.SySchedArgs)
+	if !ok {
+		return nil, fmt.Errorf("want args to be of type SySchedArgs, got %T", obj)
+	}
+
+	return SySchedArgs, nil
+}
+
+// New initializes a new plugin and returns it.
+func New(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+	sc := SySched{handle: handle}
+	sc.HostToPods = make(map[string][]*v1.Pod)
+	sc.HostSyscalls = make(map[string]map[string]bool)
+	//sc.CritSyscalls = make(map[string][]string)
```

Comment:
Usually we remove this kind of comments, or, explicitly add a TODO comment for incomplete item.